orchagent/debug_counter/drop_counter.cpp

```diff
@@ -0,0 +1,298 @@
+#include "drop_counter.h"
+
+#include "logger.h"
+#include "sai_serialize.h"
+
+using std::runtime_error;
+using std::string;
+using std::unordered_map;
+using std::unordered_set;
+using std::vector;
+
+extern sai_object_id_t gSwitchId;
+extern sai_debug_counter_api_t *sai_debug_counter_api;
+
+const int maxDropReasons = 100;
+
+// If initialization fails, this constructor will throw a runtime error.
+DropCounter::DropCounter(const string& counter_name, const string& counter_type, const unordered_set<string>& drop_reasons)
+        : DebugCounter(counter_name, counter_type), drop_reasons(drop_reasons)
+{
+    SWSS_LOG_ENTER();
+    this->initializeDropCounterInSAI();
+}
+
+DropCounter::~DropCounter()
+{
+    SWSS_LOG_ENTER();
+    try
+    {
+        DebugCounter::removeDebugCounterFromSAI();
+    }
+    catch (const std::exception& e)
+    {
+        SWSS_LOG_ERROR("Failed to remove drop counter '%s' from SAI", this->name.c_str());
```

Comment:
If exception only mean early return, why not use return code? So that you don't have to worry any caller not catching the exception?

Reply:
There are two main reasons I used exceptions instead of return codes:
1) the Drop Counter constructor can fail, and you can't use return codes in constructors, and
2) it made the code simpler in a lot of places, mostly where the drop counters are involved. Returning error codes would involve bubbling up the return code through a series of helper methods whereas throwing the exception allows us to catch it and handle it only at the top-level, which is really what we care about anyways.

I agree though that the exception behavior isn't as explicit as return codes, which can lead to issues like the comment you have above if we're not careful. I think C++ has a "noexcept" qualifier - let me see if that would help here.

Reply:
It might be helpful for us to have (just to make it easier for us to read :)) but it won't prevent this. I think the simplest solution is to move the try-catch up to doTask, but let me know if you still think that error codes are the way to go.

Reply:
I see. I am not 100% against exceptions.

I've been burnt by exception not caught within API boundaries and reaches other side of the fence. We need to be careful and explicit. I like you thoughts of helping future coders not fall into the trap.